core/base/dispatch_helper.hpp

```diff
@@ -68,52 +55,91 @@ ReturnType run_impl(std::shared_ptr<T> obj, Func&& f, Args&&... args)
- * @tparam T  the type of input object waiting converted
- * @tparam Func  the function type that is invoked if the object can be
- *               converted to pointer of const Base<K>
- * @tparam ...Args  the additional arguments for the Func
- *
- * @param obj  the input object that should be converted
- * @param f  the function will get invoked if obj can be converted successfully
- * @param args  the additional arguments for the function
+ * @note This handles the smart pointer case, i.e. either unique or shared
+ *       pointer
  */
-template <typename ReturnType, template <typename> class Base, typename K,
-          typename... Types, typename T, typename Func, typename... Args>
+template <typename ReturnType, typename K, typename... Types, typename T,
+          typename Func, typename... Args>
 ReturnType run_impl(T obj, Func&& f, Args&&... args)
 {
-    if (auto dobj = std::dynamic_pointer_cast<const Base<K>>(obj)) {
+    if (auto dobj = std::dynamic_pointer_cast<
+            with_same_constness_t<K, typename T::element_type>>(obj)) {
         return f(dobj, args...);
     } else {
-        return run_impl<ReturnType, Base, Types...>(
-            obj, std::forward<Func>(f), std::forward<Args>(args)...);
+        return run_impl<ReturnType, Types...>(obj, std::forward<Func>(f),
+                                              std::forward<Args>(args)...);
     }
 }
 
+/**
+ * Helper struct to get the result type of a function.
+ *
+ * @tparam T  Blueprint type for the function. This determines the
+ *            const-qualifier for K, as well as the pointer type
+ *            (either T*, unique_ptr<T>, or shared_ptr<T>) for K.
+ * @tparam K  The actual type to be used in the function.
+ * @tparam Func  The function to get the result from.
+ * @tparam Args  Additional arguments to the function.
+ */
+template <typename T, typename K, typename Func, typename... Args>
+struct result_of;
+
+template <typename T, typename K, typename Func, typename... Args>
+struct result_of<T*, K, Func, Args...> {
+#if __cplusplus < 201703L
+    // result_of_t is deprecated in C++17
+    using type =
+        std::result_of_t<Func(detail::with_same_constness_t<K, T>*, Args...)>;
+#else
+    using type =
+        std::invoke_result_t<Func, detail::with_same_constness_t<K, T>*,
+                             Args...>;
+#endif
+};
+
+template <typename T, typename K, typename Func, typename... Args>
+struct result_of<std::shared_ptr<T>, K, Func, Args...> {
+#if __cplusplus < 201703L
+    // result_of_t is deprecated in C++17
+    using type = std::result_of_t<Func(
+        std::shared_ptr<detail::with_same_constness_t<K, T>>, Args...)>;
+#else
+    using type = std::invoke_result_t<
+        Func, std::shared_ptr<detail::with_same_constness_t<K, T>>, Args...>;
+#endif
+};
+
+template <typename T, typename K, typename Func, typename... Args>
+struct result_of<std::unique_ptr<T>, K, Func, Args...> {
```

Comment:
the unique_ptr is unnecessary because we can not use unique_ptr in these functions